### ISSUE TITLE
WebDAV Error 401 on second visit

### ISSUE THREAD

**Issue by Coding-Kiwi** [keeweb/keeweb#1576]
**Describe the bug**

After entering the file password when using the web app for the second time, I get the error 401 instead of getting asked for the basic auth credentials.
This only happens in chrome, in firefox its fine.

**To Reproduce**
1. open keeweb, enter the file password
2. basic auth popup comes up, enter user and password
3. use keeweb as usual.
4. close browser, open browser, open keeweb, enter the file password
5. Error 401

**Expected behavior**
When the database file is received for the first time, the basic auth popup expects me to enter the username and password for webdav / basic auth. I can close and open the web app as often as I want and I only have to enter the file password and everything works. But as soon as I close the browser and open the web app again, the basic auth popup does not show up and I get error 401. I have to clear the application data using the chrome devtools, then the popup shows up again.

**Screenshots**

**Environment**
I am serving a database file via nginx / webdav with authentication enabled.
I am self hosting the web app.

KeeWeb v1.15.5 (62ea631, 2020-06-13)
Environment: web
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36

**Kdbx File**
Own database file served by webdav.

**Logs**
```
(index):1 2020-07-27T08:26:11.035Z [app] Loading config from /config.json
(index):1 2020-07-27T08:26:11.039Z [app] Loaded app config from /config.json 4ms
2020-07-27T08:26:11.181Z [updater] Service worker registered
2020-07-27T08:26:15.600Z [open:Database] File open request
(index):1 2020-07-27T08:26:15.600Z [open:Database] Open file from storage webdav
(index):1 2020-07-27T08:26:15.600Z [open:Database] Load from storage
2020-07-27T08:23:31.602Z [storage-webdav] Load /databases/database.kdbx
/databases/database.kdbx:1 Failed to load resource: the server responded with a status of 401 ()
(index):1 2020-07-27T08:23:31.627Z [storage-webdav] Load error /databases/database.kdbx 401 25ms
(index):1 2020-07-27T08:23:31.627Z [open:CodingKiwi Database] Storage load error HTTP status 401
(index):1 2020-07-27T08:23:31.627Z [open-view] Error opening file HTTP status 401
```

**Comment by antelle** [comment]
Hi! At the moment support for this kind of authentication in KeeWeb is not complete, so it has issues like this. We support credentials entered in KeeWeb, the username/password fields, but that's another thing.

**Comment by DarkCoder28** [comment]
I am also experiencing this issue with Nextcloud hosting webDAV using a username and application password for auth

**Comment by KaKi87** [comment]
Hello,

I am experiencing this issue on my self-hosted KeeWeb instance.
Both remote and local files contain recent changes.
How to proceed ?
Thanks

**Comment by rainu** [comment]
I had the same issue and debugged the code a while. I have discovered the following: After the config is loaded the files inside the config will be stored into the browser's localStorage. But the password - which is stored into the config-file - will be **not** stored into the localStorage! But the the encrypted value of the password. After you reload the page (second visit) the config-file **and** the files stored into the localStorage are loaded. But the file inside the localStorage will override the file from the config-file! Unfortunately the encrypted password will be loaded and **not** decrypted before request the vault via WebDav -> which finally causes the 401 (Unauthorized) - because of "wrong" password.

For now i am unable to discover the correct way to decrypt the password from localStorage before request via WebDav. But i discovered a workaround:

Use `"showOnlyFilesFromConfig": true` in the loaded config file. This will skip to load the stored files from localStorage.

**Comment by KaKi87** [comment]
Hello,

Whether I add this setting or not, the issue always occurs, and `encryptedPassword` is always `null`.

Thanks

**Comment by KaKi87** [comment]
I recently finally found a workaround that doesn't require emptying the app data : manually accessing the KDBX directory via normal browsing, typing the credentials, then resubmitting the password on KeeWeb.

**Comment by KaKi87** [comment]
I just now found an automated workaround : use a browser extension to add the authentication header to KDBX requests.

Here's the first one I found for that purpose : https://github.com/didierfred/SimpleModifyHeaders (tested on Chromium-based Vivaldi and Firefox-based Zen).

It doesn't work on mobile though, so I found this second one that does : https://github.com/FirefoxBar/HeaderEditor (tested on Chromium-based Kiwi for Android, available on Firefox-based Iceraven for Android).

**Comment by Aetherinox** [comment]
We are currently working on updating the WebDAV functionality (including the ability to authenticate with Digests, Basic, etc.

**Comment by KaKi87** [comment]
> can you confirm the issue still exists in 1.18.7?

Yes.

> I just tried OP's steps in 1.19.x and I don't get this issue, it opens every time without a 401.

Do you have basic authentication on your WebDAV ?

**Comment by Aetherinox** [comment]
> Do you have basic authentication on your WebDAV ?

Yes, Basic Auth is the only authentication method I have right now. However, I am using IIS + Webdav and handling reverse proxy through Traefik. So my setup differs from OP (nginx) and I'm sure probably you as well.

What setup of WebDAV are you on.

**Comment by KaKi87** [comment]
How exotic. I'm classical : Apache on Linux.

**Comment by Aetherinox** [comment]
A little over-kill for me right now, but I have a Linux box I also use for testing. I'll take some time tonight and configure my Linux box to run webdav. I need to do so anyway so that I have multiple test environments. The IIS + Traefik setup was a throw-together. 90% of the time, I run on Linux.

**Edit**: I remember why I went the IIS route, I was going to remote into my box the other night, and the host was updating something at the datacenter, so my connection was down and I didn't want to wait.
